merkle/hashers/tree_hasher.go

```diff
@@ -33,6 +35,32 @@ type LogHasher interface {
 	Size() int
 }
 
+// InplaceLogHasher provides the hash functions needed to compute dense merkle
+// trees.
+// Implementations may hold state and are not expected to be thread
+// safe. They should not be shared between multiple callers. Where performance
+// is not critical the LogHasher interface provides a safer alternative.
+//
+// Note: Implementations are not expected to be thread safe. The caller must
+// ensure that concurrent calls are not made and is responsible for making
+// copies of slices where necessary to avoid data being overwritten.
+type InplaceLogHasher interface {
```

Comment:
Maybe say somewhere that capacity of `res` slices should be at least `Size()`?

Reply:
They don't have to be. Tests added for that case.

Reply:
Ah, indeed, Go can re-allocate slices on `append` under the hood.